src/bgw/scheduler.c

```diff
@@ -0,0 +1,386 @@
+ *
+ * Limitations: For now the jobs are only loaded when the scheduler starts and are not
+ * updated if the jobs table changes
+ *
+ * TODO: right now jobs are not prioritized when slots available. Should prioritize in
+ * asc next_start order.
+ *
+*/
+#include <postgres.h>
+
+#include <miscadmin.h>
+#include <postmaster/bgworker.h>
+#include <storage/ipc.h>
+#include <storage/latch.h>
+#include <storage/lwlock.h>
+#include <storage/proc.h>
+#include <storage/shmem.h>
+#include <utils/jsonb.h>
+#include <utils/timestamp.h>
+#include <utils/snapmgr.h>
+#include <utils/memutils.h>
+#include <access/xact.h>
+#include <pgstat.h>
+
+#include "extension.h"
+#include "scheduler.h"
+#include "job.h"
+#include "job_stat.h"
+#include "version.h"
+#include "compat.h"
+#include "timer.h"
+
+#define SCHEDULER_APPNAME "TimescaleDB Background Worker Scheduler"
+
+#define LEAST_TIMESTAMP(left, right) \
+	(left < right ? left : right)
+
+TS_FUNCTION_INFO_V1(ts_bgw_scheduler_main);
+
+
+/* has to be global to shutdown jobs on exit */
+static List *scheduled_jobs = NIL;
+
+typedef enum JobState
+{
+	/* terminal state */
+	JOB_STATE_DISABLED,
+
+	/* next states: JOB_STATE_STARTED */
+	JOB_STATE_WAITING,
+
+	/*
+	 * next states: JOB_STATE_WAITING, JOB_STATE_DISABLED,
+	 * JOB_STATE_TERMINATING
+	 */
+	JOB_STATE_STARTED,
+
+	/* next states: JOB_STATE_WAITING, JOB_STATE_DISABLED */
+	JOB_STATE_TERMINATING,
+} JobState;
+
+typedef struct ScheduledBgwJob
+{
+	BgwJob		job;
+	TimestampTz next_start;
+	TimestampTz timeout_at;
+	JobState	state;
+	BackgroundWorkerHandle *handle;
+} ScheduledBgwJob;
+
+static void scheduled_bgw_job_mark_stopped(ScheduledBgwJob *job);
+
+static inline void
+bgw_scheduler_on_postmaster_death(void)
+{
+	/*
+	 * Don't call exit hooks cause we want to bail out quickly. We don't care
+	 * about cleaning up shared memory in this case anyway since it's
+	 * potentially corrupt.
+	 */
+	on_exit_reset();
+	ereport(FATAL,
+			(errcode(ERRCODE_ADMIN_SHUTDOWN),
+			 errmsg("postmaster exited while TimescaleDB scheduler was working")));
+}
+
+/*
+*	This function starts a job.
+*	To correctly count crashes we need to mark the start of a job in a separate
+*	txn before we kick off the actual job. Thus this function cannot be run
+*	from within a transaction.
+*/
+static void
+scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callback_type bgw_register)
+{
+	pid_t		pid;
+	BgwHandleStatus status;
+
+
+	Assert(sjob->state == JOB_STATE_WAITING);
+
+	StartTransactionCommand();
+	bgw_job_stat_mark_start(sjob->job.fd.id);
+	if (bgw_job_has_timeout(&sjob->job))
+		sjob->timeout_at = bgw_job_timeout_at(&sjob->job, timer_get_current_timestamp());
+	else
+		sjob->timeout_at = DT_NOEND;
+	CommitTransactionCommand();
+
+	elog(DEBUG1, "scheduling job %d \"%s\"", sjob->job.fd.id, NameStr(sjob->job.fd.application_name));
+
+
+	sjob->handle = bgw_job_start(&sjob->job);
+	sjob->state = JOB_STATE_STARTED;
```

Comment:
Shouldn't there be a single function to transition to `JOB_STATE_STARTED`, which also sets the state? I.e., this state should be set in `bgw_job_start()`, no?

Reply:
Fixed